PumaGrasshopper/ComponentPumaBulkInput.cs

```diff
@@ -0,0 +1,144 @@
+﻿/**
+ * Puma - CityEngine Plugin for Rhinoceros
+ *
+ * See https://esri.github.io/cityengine/puma for documentation.
+ *
+ * Copyright (c) 2021-2023 Esri R&D Center Zurich
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Grasshopper.Kernel;
+using Grasshopper.Kernel.Data;
+using Grasshopper.Kernel.Parameters;
+using Grasshopper.Kernel.Types;
+using PumaGrasshopper.Properties;
+using Rhino.Geometry;
+
+namespace PumaGrasshopper
+{
+    
+    public class ComponentPumaBulkInput : ComponentPumaShared
+    {
+        const string BULK_INPUT_NAME = "Rule attributes";
+        const string BULK_INPUT_NICK_NAME = "Attributes";
+        const string BULK_INPUT_DESC = "Attributes passed to the rule. Expects a tree of key:value pairs.";
```

Comment:
Isn't it more a list/map of key:value pairs from the users perspective?

Reply:
In Grasshopper, the usual data structure is called a Tree. A Tree can have several branches, called paths. In this case, each path would be a list of key:value pairs applied to one input shape. So the list at path 0 is applied to the input shape at path 0, and so on.